In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
metadata_df = pd.read_csv("C:/Users/abhij/Downloads/metadata.csv")

In [3]:
metadata_df.head(5)


Record ID                                          File Name Assignment ID  \
0  207431327         HearstTransportation_CAMP_MontrealTeam.png      CPR6931p   
1  207431332  HearstTransportation_MicrosoftTeams_ZoomMeetin...      CPR6931p   
2  207431348   HearstTransportation_BlackBook_LauraWehunt_1.png      CPR6931p   
3  207431356   HearstTransportation_BlackBook_LauraWehunt_2.png      CPR6931p   
4  207432753                   HearstMagazines_CDS_Employee.png      CPR6931p   

  Year                                             Person  \
0  NaN                                       Unidentified   
1  NaN  Unidentified,Don Dickson,Danya Bynoe,Varun Kum...   
2  NaN                                       Laura Wehunt   
3  NaN                                       Laura Wehunt   
4  NaN                                     Bryan Phillips   

                                            Keywords Creators Date  
0  People @ Work,Hearst transportation,Transporta...      NaN  NaN  
1  People @ Work,Hearst transportation,Transporta...      NaN  NaN  
2  People @ Work,Hearst transportation,Headshots,...      NaN  NaN  
3  People @ Work,Hearst transportation,Headshots,...      NaN  NaN  
4  People @ Work,CDS,Hearst Magazines,Hearst maga...      NaN  NaN

In [4]:
metadata_df.groupby(by = ['Keywords']).agg('count').sort_values(by = ['Record ID'], ascending=False).head(30)

Record ID  File Name  \
Keywords                                                                   
People @ Work,Homecoming,2013,Events,44th Floor...        443        443   
People @ Work,2018,Events,Friday,Texas,Texas 20...        258        258   
Archdiocese of New York,Corporate,Al Smith Dinn...        230        230   
People @ Work,Spotlight 2016,Awards,Hearst Maga...        140        140   
People @ Work,Frank A. Bennack,Frank Bennack,Ev...        127        127   
People @ Work,2018,Thursday,Events,Texas,Texas ...        114        114   
People @ Work,Tower Gala,Events,Hearst Tower              111        111   
People @ Work,2019,Events,Friday,Texas,Texas 2019          98         98   
People @ Work,Texas 2016,2016,Friday,Events,Texas          97         97   
People @ Work,Texas 2017,2017,Events,Friday,Texas          96         96   
Scenery,Hearst Real Estate,2021,Outdoors,San Si...         94         94   
People @ Work,Culture Shifting Weekend,Culture ...         90         90   
People @ Work,2019,Leave Something on the Table...         89         89   
44th Floor,Events,Hearst Ventures,Ventures,Ken ...         86         86   
People @ Work,2019,Thursday,Events,Texas,Texas ...         76         76   
Hearst Real Estate,2021,Outdoors,San Simeon,Hor...         73         73   
People @ Work,2019,Awards,Hearst Magazines,Even...         72         72   
Cooking,Jack Ranch,Hearst Real Estate,2021,Hear...         68         68   
People @ Work,Texas 2016,2016,Thursday,Events,T...         66         66   
People @ Work,Texas 2016,Alison Krauss,2016,Fri...         65         65   
People @ Work,2018,Awards,Hearst Magazines,Even...         64         64   
People @ Work,2017,Awards,Hearst Magazines,Even...         62         62   
People @ Work,2018,Frank A. Bennack,Thursday,Fr...         61         61   
People @ Work,Editorial Excellence,Hearst Magaz...         60         60   
Frank Bennack,FAB,Homecoming,44th Floor,People ...         60         60   
People @ Work,Owl's Nest,Event,44th Floor                  59         59   
People @ Work,Texas 2017,2017,Thursday,Events,T...         58         58   
People @ Work,Editorial Excellence,2018,Hearst ...         56         56   
People @ Work,Texas 2017,2017,Frank A. Bennack,...         55         55   
Hearst.com,Headshots,1276x960,Portraits,HearstL...         54         54   

                                                    Assignment ID  Year  \
Keywords                                                                  
People @ Work,Homecoming,2013,Events,44th Floor...            443   443   
People @ Work,2018,Events,Friday,Texas,Texas 20...            258   258   
Archdiocese of New York,Corporate,Al Smith Dinn...              0   230   
People @ Work,Spotlight 2016,Awards,Hearst Maga...              0   140   
People @ Work,Frank A. Bennack,Frank Bennack,Ev...            127   127   
People @ Work,2018,Thursday,Events,Texas,Texas ...            114   114   
People @ Work,Tower Gala,Events,Hearst Tower                  111   111   
People @ Work,2019,Events,Friday,Texas,Texas 2019              98    98   
People @ Work,Texas 2016,2016,Friday,Events,Texas              97    97   
People @ Work,Texas 2017,2017,Events,Friday,Texas              96    96   
Scenery,Hearst Real Estate,2021,Outdoors,San Si...             94    94   
People @ Work,Culture Shifting Weekend,Culture ...             89    90   
People @ Work,2019,Leave Something on the Table...              0    89   
44th Floor,Events,Hearst Ventures,Ventures,Ken ...              0    86   
People @ Work,2019,Thursday,Events,Texas,Texas ...             76    76   
Hearst Real Estate,2021,Outdoors,San Simeon,Hor...             73    73   
People @ Work,2019,Awards,Hearst Magazines,Even...              0    72   
Cooking,Jack Ranch,Hearst Real Estate,2021,Hear...             68    68   
People @ Work,Texas 2016,2016,Thursday,Events,T...             66    66   
People @ Work,Texas 2016,Alison Krauss,2016,

## Feature Engineering

#### Split on comma and find individual keywords

In [5]:
clean_df = metadata_df.copy()

In [6]:
clean_df = clean_df.join(clean_df['Keywords'].str.get_dummies(','))

In [7]:
tag_df = pd.DataFrame(clean_df.iloc[:,7:].sum(axis = 0), columns = ['Count']).sort_values(by = ['Count'], ascending = False)

In [8]:
tag_df['Tag_Name'] = tag_df.index
tag_df.reset_index(drop = True, inplace = True)

In [9]:
### Further steps
tag_df['Count_tag_appears'] = tag_df['Count']
tag_df = tag_df.iloc[:,1:]

In [10]:
tag_df

Tag_Name  Count_tag_appears
0        People @ Work               7058
1               Events               6483
2           44th Floor               2993
3                Texas               2090
4     Hearst Magazines               1611
...                ...                ...
1446   Joanne Alfarone                  1
1447      John Francis                  1
1448           Johnson                  1
1449   Jonathan Becker                  1
1450         Portraits                  1

[1451 rows x 2 columns]

In [13]:
tag_df[tag_df['Count_tag_appears']>100].count()

Tag_Name             116
Count_tag_appears    116
dtype: int64

#### Finding A Subset of Tags
* There are 116 tags that have a count of atleast 100 in the dataset
*Find all tags that have a count of atleast 100. Set this as an assorted list of tags
*Check in clean_df the count of images where any one tag appears which is in the assorted list of tags
*Remove year of picture From this tag list.
*Ultimately reduce tags created as dummy features in original dataset to reduce dimensionality

In [46]:
tag_list = list(tag_df[tag_df['Count_tag_appears']>140]['Tag_Name'])

In [54]:
#Count of images that can be excluded cannot be more than 0.5% of the dataset size. 0.5% of 9300 images roughly is 46.5. On Trial and Error a cutoff of 140 tags is found.
clean_df[(clean_df[tag_list].any(axis = 1))==False].count()

Record ID         45
File Name         45
Assignment ID     34
Year              12
Person            37
                  ..
workout           45
  Gil Maurer      45
 Headshots        45
 People @ Work    45
 Portraits        45
Length: 1459, dtype: int64

In [52]:
#Number of features to be used in dataset for training before feature importance is taken
len(tag_list)+ metadata_df.shape[1] - 1

103

In [56]:
image_df_original = clean_df.iloc[:,[0,1,2,3,4,6,7]]

In [60]:
image_df_original = image_df_original.join(clean_df[tag_list])

In [62]:
# Final Dataset for Further Modelling

In [61]:
image_df_original

Record ID                                          File Name  \
0     207431327         HearstTransportation_CAMP_MontrealTeam.png   
1     207431332  HearstTransportation_MicrosoftTeams_ZoomMeetin...   
2     207431348   HearstTransportation_BlackBook_LauraWehunt_1.png   
3     207431356   HearstTransportation_BlackBook_LauraWehunt_2.png   
4     207432753                   HearstMagazines_CDS_Employee.png   
...         ...                                                ...   
9294  206293007                      SteveSwartzHeadshot-FINAL.jpg   
9295  206293008                              SteveSwartz2-2013.jpg   
9296  206293009                             SteveSwartz3-2013-.jpg   
9297  206293010                SwartzJames DuffNewseum03142013.jpg   
9298  206293011                              SteveSwartz6-2013.jpg   

     Assignment ID        Year  \
0         CPR6931p         NaN   
1         CPR6931p         NaN   
2         CPR6931p         NaN   
3         CPR6931p         NaN   
4         CPR6931p         NaN   
...            ...         ...   
9294           NaN  2012-12-13   
9295           NaN  2013-01-22   
9296           NaN  2013-01-22   
9297           NaN  2013-03-14   
9298           NaN  2013-01-22   

                                                 Person  \
0                                          Unidentified   
1     Unidentified,Don Dickson,Danya Bynoe,Varun Kum...   
2                                          Laura Wehunt   
3                                          Laura Wehunt   
4                                        Bryan Phillips   
...                                                 ...   
9294                                       Steve Swartz   
9295                                       Steve Swartz   
9296                                       Steve Swartz   
9297                            Steve Swartz,James Duff   
9298                                       Steve Swartz   

                      Creators Date  People @ Work  Events  44th Floor  ...  \
0                          NaN  NaN              1       0           0  ...   
1                          NaN  NaN              1       0           0  ...   
2                          NaN  NaN              1       0           0  ...   
3                          NaN  NaN              1       0           0  ...   
4                          NaN  NaN              1       0           0  ...   
...                        ...  ...            ...     ...         ...  ...   
9294                       NaN  NaN              0       0           0  ...   
9295                       NaN  NaN              0       0           0  ...   
9296  PHILIP FRIEDMAN/Studio D  NaN              0       0           0  ...   
9297               Paul Morigi  NaN              0       0           0  ...   
9298  PHILIP FRIEDMAN/Studio D  NaN              0       0           0  ...   

      HearstLab  Kate Lewis  David Carey  Colin Powell  Norman Foster  \
0             0           0            0             0              0   
1             0           0            0             0              0   
2             0           0            0             0              0   
3             0           0            0             0              0   
4             0           0            0             0              0   
...         ...         ...          ...           ...            ...   
9294          0           0            0             0              0   
9295          0           0            0             0              0   
9296          0           0            0             0              0   
9297          0           0            0             0              0   
9298          0           0            0             0              0   

      Culture Shift Labs  Hearst Health  Gathering  Horseback Riding  \
0                      0              0          0                 0   
1                      0              0          0                 0   
2    